### ISSUE TITLE
Error while importing

### ISSUE THREAD

**Issue by AxisNL** [JC5/firefly-iii#249]
Hmm.. trying to import a CSV with two transactions, I get another error. Perhaps something is wrong with my installation. Do you see any correlation between the three issues I just reported? Perhaps some cache should be cleared?

Error

This error occured in file /home/domains/hongens/firefly/app/Helpers/Csv/PostProcessing/Bill.php on line 23 with code 0.

Stack trace

```
#0 /home/domains/hongens/firefly/app/Helpers/Csv/PostProcessing/Bill.php(23): Illuminate\Foundation\Bootstrap\HandleExceptions->handleError(8, 'Trying to get p...', '/home/domains/h...', 23, Array)
#1 /home/domains/hongens/firefly/app/Helpers/Csv/Importer.php(299): FireflyIII\Helpers\Csv\PostProcessing\Bill->process()
#2 /home/domains/hongens/firefly/app/Helpers/Csv/Importer.php(250): FireflyIII\Helpers\Csv\Importer->postProcess()
#3 /home/domains/hongens/firefly/app/Helpers/Csv/Importer.php(114): FireflyIII\Helpers\Csv\Importer->importRow()
#4 /home/domains/hongens/firefly/app/Http/Controllers/CsvController.php(319): FireflyIII\Helpers\Csv\Importer->run()
#5 [internal function]: FireflyIII\Http\Controllers\CsvController->process()
#6 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9268): call_user_func_array(Array, Array)
#7 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9330): Illuminate\Routing\Controller->callAction('process', Array)
#8 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9310): Illuminate\Routing\ControllerDispatcher->call(Object(FireflyIII\Http\Controllers\CsvController), Object(Illuminate\Routing\Route), 'process')
#9 [internal function]: Illuminate\Routing\ControllerDispatcher->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#10 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(52): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#11 [internal function]: Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#12 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9787): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#13 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9311): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#14 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9298): Illuminate\Routing\ControllerDispatcher->callWithinStack(Object(FireflyIII\Http\Controllers\CsvController), Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request), 'process')
#15 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8385): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request), 'FireflyIII\\Http...', 'process')
#16 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8372): Illuminate\Routing\Route->runController(Object(Illuminate\Http\Request))
#17 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8086): Illuminate\Routing\Route->run(Object(Illuminate\Http\Request))
#18 [internal function]: Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#19 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(52): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#20 /home/domains/hongens/firefly/app/Http/Middleware/Binder.php(45): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#21 [internal function]: FireflyIII\Http\Middleware\Binder->handle(Object(Illuminate\Http\Request), Object(Closure))
#22 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#23 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#24 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#25 /home/domains/hongens/firefly/app/Http/Middleware/Range.php(90): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#26 [internal function]: FireflyIII\Http\Middleware\Range->handle(Object(Illuminate\Http\Request), Object(Closure))
#27 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#28 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#29 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#30 /home/domains/hongens/firefly/app/Http/Middleware/IsConfirmed.php(59): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#31 [internal function]: FireflyIII\Http\Middleware\IsConfirmed->handle(Object(Illuminate\Http\Request), Object(Closure))
#32 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#33 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#34 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#35 /home/domains/hongens/firefly/app/Http/Middleware/AuthenticateTwoFactor.php(63): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#36 [internal function]: FireflyIII\Http\Middleware\AuthenticateTwoFactor->handle(Object(Illuminate\Http\Request), Object(Closure))
#37 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#38 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#39 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#40 /home/domains/hongens/firefly/app/Http/Middleware/Authenticate.php(44): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#41 [internal function]: FireflyIII\Http\Middleware\Authenticate->handle(Object(Illuminate\Http\Request), Object(Closure))
#42 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#43 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#44 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#45 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(3165): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#46 [internal function]: Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle(Object(Illuminate\Http\Request), Object(Closure))
#47 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#48 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#49 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#50 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(13221): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#51 [internal function]: Illuminate\View\Middleware\ShareErrorsFromSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#52 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#53 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#54 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#55 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(11781): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#56 [internal function]: Illuminate\Session\Middleware\StartSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#57 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#58 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#59 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#60 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(12960): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#61 [internal function]: Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle(Object(Illuminate\Http\Request), Object(Closure))
#62 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#63 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#64 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#65 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(12897): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#66 [internal function]: Illuminate\Cookie\Middleware\EncryptCookies->handle(Object(Illuminate\Http\Request), Object(Closure))
#67 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#68 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#69 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#70 [internal function]: Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#71 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9787): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#72 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8087): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#73 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8078): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#74 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8068): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#75 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(2393): Illuminate\Routing\Router->dispatch(Object(Illuminate\Http\Request))
#76 [internal function]: Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#77 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(52): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#78 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(3226): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#79 [internal function]: Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(Object(Illuminate\Http\Request), Object(Closure))
#80 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#81 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#82 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#83 [internal function]: Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#84 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9787): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#85 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(2340): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#86 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(2324): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#87 /home/domains/hongens/firefly/public/index.php(53): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#88 {main}
```

**Comment by JC5** [comment]
You've indicated one of your fields to be the "bill" field (bill name? bill id?).

Do those bills exist already?

**Comment by AxisNL** [comment]
Nope, I did not configure one of the bill fields.. 

I'm using the config file I have been using successfully for months: { "date-format": "Ymd", "has-headers": false, "map": [], "roles": { "0": "account-iban", "1": "currency-code", "2": "date-transaction", "3": "rabo-debet-credit", "4": "amount", "5": "opposing-iban", "6": "opposing-name", "10": "description", "11": "description" }, "mapped": [], "specifix": [ "RabobankDescription" ] }

But since I see "#0 ../Helpers/Csv/PostProcessing/Bill.php(23): ..", could it be possible that after importing it's trying to apply bills and that it gets an error there?

**Comment by JC5** [comment]
Mmm, that is quite possible. I'll look into it today or tomorrow.

With the arrival of PHP7, I've forced most methods in Firefly to always return the same type of object. Returning "null", or nothing, is no longer allowed. Instead, an empty object is returned (which is a slightly different thing).

A lot of code still checks for these "null" values instead of for empty objects. That might cause this bug.

**Comment by JC5** [comment]
@Bonno has fixed some things in the bill processor. Could that have fixed this issue?

**Comment by Bonno** [comment]
Yes, this is the same problem that was fixed in 99c219e

**Comment by JC5** [comment]
I'm closing the issue now, because I think it's fixed. The CSV importer is due for an overhaul. Let me know if problems occur.
